src/web-auth/index.js

```diff
@@ -162,37 +159,62 @@ WebAuth.prototype.parseHash = function(options, cb) {
   ) {
     return cb(null, null);
   }
+  return this.validateAuthenticationResponse(options, parsedQs, cb);
+};
 
-  state = parsedQs.state || options.state;
-
-  transaction = this.transactionManager.getStoredTransaction(state);
-  transactionNonce = options.nonce || (transaction && transaction.nonce) || null;
+/**
+ * Validates an Auth response from a Auth flow started with {@link authorize}
+ *
+ * Only validates id_tokens signed by Auth0 using the RS256 algorithm using the public key exposed
+ * by the `/.well-known/jwks.json` endpoint of your account.
+ * Tokens signed with other algorithms, e.g. HS256 will not be accepted.
+ *
+ * @method validateAuthenticationResponse
+ * @param {Object} options
+ * @param {String} options.hash the url hash. If not provided it will extract from window.location.hash
+ * @param {String} [options.state] value originally sent in `state` parameter to {@link authorize} to mitigate XSRF
+ * @param {String} [options.nonce] value originally sent in `nonce` parameter to {@link authorize} to prevent replay attacks
+ * @param {String} [options._idTokenVerification] makes parseHash perform or skip `id_token` verification. We **strongly** recommend validating the `id_token` yourself if you disable the verification.
+ * @param {authorizeCallback} cb
+ */
+WebAuth.prototype.validateAuthenticationResponse = function(options, parsedHash, cb) {
+  var state = parsedHash.state || options.state || this.baseOptions.state;
+  var transaction = this.transactionManager.getStoredTransaction(state);
+  var transactionStateMatchesState = transaction && transaction.state === state;
+  if (state && !transactionStateMatchesState) {
+    return cb({
+      error: 'invalid_token',
+      errorDescription: '`state` does not match.'
+    });
+  }
+  var transactionNonce =
+    options.nonce || (transaction && transaction.nonce) || this.baseOptions.nonce || null;
```

Comment:
if you explicitly provide a nonce, we check that instead. if not, we check the transaction. if there's no transaction, we check if you passed a nonce when you created the webauth object. if not, we set it to null because that's what the idtoken-verifier expects

Reply:
We should only be checking against `transaction.nonce`, as that's were we kept the "state of the auth" (state/nonce) in the first place.

Reply:
won't do that right now. tracking here: https://github.com/auth0/auth0.js/issues/530